File: BaekJoon/Problems/Step/09/9506.cpp
// 9506. Federation Favorites(약수들의 합)
// 2023.07.02

/* Sample input & output data
6
12
28
-1

6 = 1 + 2 + 3
12 is NOT perfect.
28 = 1 + 2 + 4 + 7 + 14
*/


#include <iostream>
#include <vector>

// #define test
#define endl '\n'

using namespace std;

int main()
{
    while (true)
    {
        // Input
        int n;
        cin >> n;
        if (n == -1) break;

        // Operate
        vector<int> v;
        int sum = 0;
        for (int i = 1; i <= n/2; i++)
        {
            if (n % i == 0)
            {
                v.push_back(i);
                sum += i;
            }

            if (n < sum) break;
        }

        // Output
        if (sum == n)
        {
            cout << n << " = 1";
            for (int i = 1; i < v.size(); i++) cout << " + " << v[i];
        }
        else cout << n << " is NOT perfect.";
        cout << endl;
    }

    return 0;
}

File: BaekJoon/Problems/Step/09/9506.rs
// 9506. Federation Favorites (약수들의 합)
// 2024.04.05


use std::io;

fn main()
{
    // let test: bool = true;
    let test: bool = false;

    loop
    {
        // Input n
        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap();
        let n: i32 = input.trim().parse().unwrap();

        if n == -1 { break; }
        else
        {
            // Calc.
            let mut sum: i32 = 1;
            let mut str: String = format!("{} = 1", n);
            for i in 2..=n/2
            {
                if n % i == 0
                {
                    sum += i;
                    str += &format!(" + {}", i);
                }
                // Test
                if test { println!("{} : {} {}", i, sum, n); }

                if sum > n { break; }
            }

            // Output
            if sum == n { println!("{}", str); }
            else { println!("{} is NOT perfect.", n)}
        }
    }
}
